lib/WebpackOptionsDefaulter.js

```diff
@@ -113,6 +113,32 @@ class WebpackOptionsDefaulter extends OptionsDefaulter {
 			}
 			return "[id].js";
 		});
+		this.set("output.moduleFilename", "make", options => {
```

Comment:
Yes, either a 'top-level'  `output` option for each type (for `moduleFilename` (`${type}moduleFilename`)) is still needed here (e.g to support specifying different subdirectories per type like `images/[modulehash].[ext]`) or the `output` 'top-level' gets grouped per type (maybe a good idea in regards to HTML/CSS/* Modules in the future, but this can be addressed later (also not sure about it))

**`webpack.config.js`**
```js
const config = {
   output: {
       js: {
         filename: '[id].[name].[hash].[ext]',
         chunkfilename: '[id].[name].[hash].[ext]'
      },
      css: {
         filename: '[id].[name].[hash].[ext]',
         chunkfilename: '[id].[name].[hash].[ext]',
         moduleFilename: '[id].[name].[hash].[ext]',
      },
      url: {
         filename: '[id].[name].[hash].[ext]' // used as moduleFilename
         modulefilename: '[id].[name].[hash].[ext]',
      }
   }
}
```

But this way it's also not 💯 sound e.g `output.(url|wasm).filename` && especially `output.(url|wasm).chunkFilname` wouldn't make any sense at all

Reply:
Without `test/includes/exclude` it is doesn't make sense, a lot of people use multiple `url/file-loader` with difference `test/include/exclude`. Maybe better:

```js
const config = {
   output: {
      {
         test: /\.(png|gif|svg)$/i,
         include: /src/,
         filename: 'app/[id].[name].[hash].[ext]',
         chunkfilename: 'app/[id].[name].[hash].[ext]'
      },
      {
         test: /\.(png|gif|svg)$/i,
         include: /node_modules/,
         filename: 'vendor/[id].[name].[hash].[ext]',
         chunkfilename: 'vendor/[id].[name].[hash].[ext]'
      }
   }
}
```

Same for `js/css/etc`

Reply:
Would `test|include|exclude` not still be handled by `module.rules` (those apply at a different phase/stage)? The basic idea is 

```js
output: {
   [type]: {
     // 1. Common options (filename[chunkFilename [, moduleFilename]])
     // 2. Common placeholders (`[id]`, `[name]`, `[hash]`, `[ext]`)
     // e.g 
     filename: '[name].[hash].[ext]'
   },
   ...types
}
```

But this needs more dicussion

Reply:
@michael-ciniawsky basic idea not cover when you want to move difference assets to difference directories (example - assets from `node_modules` to `build/vendor/assets` and assets from `app` to `build/app/assets`), many people do this, i don't know why, but they do.